src/main/java/org/truffleruby/language/globals/ReadMatchReferenceNodes.java

```diff
@@ -0,0 +1,315 @@
+
+        public ReadMatchNode(GetFromThreadAndFrameLocalStorageNode readMatchNode) {
+            this.readMatchNode = readMatchNode;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            final Object match = readMatchNode.execute(frame);
+
+            if (matchNilProfile.profile(match == nil())) {
+                return nil();
+            }
+
+            final DynamicObject matchData = (DynamicObject) match;
+
+            return tosMatchNode.execute(matchData);
+        }
+
+        @Override
+        public Object isDefined(VirtualFrame frame) {
+            if (isNil(execute(frame))) {
+                return nil();
+            } else {
+                return coreStrings().GLOBAL_VARIABLE.createInstance();
+            }
+        }
+    }
+
+    public static class ReadHighestMatchNode extends RubyNode {
+        @Child private GetFromThreadAndFrameLocalStorageNode readMatchNode;
+        @Child private ValuesNode getValues = ValuesNode.create();
+
+        protected final ConditionProfile matchNilProfile = ConditionProfile.createBinaryProfile();
+
+        public ReadHighestMatchNode(GetFromThreadAndFrameLocalStorageNode readMatchNode) {
+            this.readMatchNode = readMatchNode;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            final Object match = readMatchNode.execute(frame);
+
+            if (matchNilProfile.profile(match == nil())) {
+                return nil();
+            }
+
+            final DynamicObject matchData = (DynamicObject) match;
+
+            final Object[] values = getValues.execute(matchData);
+
+            for (int n = values.length - 1; n >= 0; n--)
+                if (values[n] != nil()) {
+                    return values[n];
+                }
+
+            return nil();
+        }
+
+        @Override
+        public Object isDefined(VirtualFrame frame) {
+            if (isNil(execute(frame))) {
+                return nil();
+            } else {
+                return coreStrings().GLOBAL_VARIABLE.createInstance();
+            }
+        }
+    }
+
+    public static class ReadNthMatchNode extends RubyNode {
+        @Child private GetFromThreadAndFrameLocalStorageNode readMatchNode;
+        @Child private ValuesNode getValues = ValuesNode.create();
+        private final int index;
+
+        protected final ConditionProfile matchNilProfile = ConditionProfile.createBinaryProfile();
+
+        public ReadNthMatchNode(GetFromThreadAndFrameLocalStorageNode readMatchNode, int index) {
+            this.readMatchNode = readMatchNode;
+            this.index = index;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            final Object match = readMatchNode.execute(frame);
+
+            if (matchNilProfile.profile(match == nil())) {
+                return nil();
+            }
+
+            final DynamicObject matchData = (DynamicObject) match;
```

Comment:
Why cast this here instead of the first assignment?
`final DynamicObject matchData = (DynamicObject) readMatchNode.execute(frame);`